File: 알고리즘 스터디 문제들/Part 7 - 백트래킹/문제/연산자 끼워넣기.cpp
#include <iostream>
#include <array>
using namespace std;

int N;
int max_val = INT32_MIN;
int min_val = INT32_MAX;

array<int, 11> num;
array<int, 4> oper; // 연산자의 개수

void backtrack(int cnt, int total)
{
    if (cnt == N - 1)
    {
        max_val = max(max_val, total);
        min_val = min(min_val, total);
        return;
    }

    if (oper[0] > 0)
    {
        oper[0]--;
        backtrack(cnt + 1, total + num[cnt + 1]);
        oper[0]++;
    }

    if (oper[1] > 0)
    {
        oper[1]--;
        backtrack(cnt + 1, total - num[cnt + 1]);
        oper[1]++;
    }

    if (oper[2] > 0)
    {
        oper[2]--;
        backtrack(cnt + 1, total * num[cnt + 1]);
        oper[2]++;
    }

    if (oper[3] > 0)
    {
        oper[3]--;
        backtrack(cnt + 1, total / num[cnt + 1]);
        oper[3]++;
    }
}

int main()
{
    cin >> N;
    for (size_t i = 0; i < N; i++)
        cin >> num[i];
    for (size_t i = 0; i < 4; i++)
        cin >> oper[i];

    backtrack(0, num[0]);
    cout << max_val << '\n'
         << min_val;
}

File: 알고리즘 스터디 문제들/Part 7 - 백트래킹/문제/연산자 끼워넣기.py
def backtrack(cur, total):
    global nums, oper, min_val, max_val, N
    if (cur == N - 1):
        max_val = max(max_val, total)
        min_val = min(min_val, total)
        return

    if (oper[0] > 0):
        oper[0] -= 1
        backtrack(cur + 1, total + nums[cur + 1])
        oper[0] += 1

    if (oper[1] > 0):
        oper[1] -= 1
        backtrack(cur + 1, total - nums[cur + 1])
        oper[1] += 1

    if (oper[2] > 0):
        oper[2] -= 1
        backtrack(cur + 1, total * nums[cur + 1])
        oper[2] += 1

    if (oper[3] > 0):
        oper[3] -= 1

        # Python과 C++의 나눗셈 연산자는 값이 다르다.
        backtrack(cur + 1, int(total / nums[cur + 1]))
        oper[3] += 1


N = int(input())
nums = list(map(int, input().split()))
oper = list(map(int, input().split()))
max_val = -1100
min_val = 100 ** 11

backtrack(0, nums[0])
print(max_val, min_val, sep="\n")
